match/post_ohr.py

```diff
@@ -0,0 +1,333 @@
+
+    matcher = ThresholdMatcher(
+        ColumnsIndex(["fc"]),
+        {
+            "first_name": JaroWinklerSimilarity(),
+            "last_name": JaroWinklerSimilarity(),
+            "agency": JaroWinklerSimilarity(),
+        },
+        dfa,
+        dfb,
+    )
+    decision = 0.950
+
+    matches = matcher.get_index_pairs_within_thresholds(decision)
+    match_dict = dict(matches)
+
+    # value is added to "extracted_uid" column if match score is 1
+    cprr.loc[:, "matched_uid_7"] = cprr.uid_7.map(lambda x: match_dict.get(x))
+    return cprr
+
+
+def match_uid_8_and_personnel(cprr, pprr):
+    dfa = cprr[["uid_8", "first_name", "last_name", "agency_8", "middle_name"]]
+    dfa = dfa.rename(columns={"agency_8": "agency"})
+    dfa = dfa.drop_duplicates(subset=["uid_8"]).set_index("uid_8")
+    dfa.loc[:, "fc"] = dfa.first_name.fillna("").map(lambda x: x[:1])
+    dfa.loc[:, "lc"] = dfa.last_name.fillna("").map(lambda x: x[:1])
+
+    dfb = pprr[["uid", "first_name", "last_name", "agency", "middle_name"]]
+    dfb = dfb.drop_duplicates(subset=["uid"]).set_index("uid")
+    dfb.loc[:, "fc"] = dfb.first_name.fillna("").map(lambda x: x[:1])
+    dfb.loc[:, "lc"] = dfb.last_name.fillna("").map(lambda x: x[:1])
+
+    matcher = ThresholdMatcher(
+        ColumnsIndex(["fc"]),
+        {
+            "first_name": JaroWinklerSimilarity(),
+            "last_name": JaroWinklerSimilarity(),
+            "agency": JaroWinklerSimilarity(),
+        },
+        dfa,
+        dfb,
+    )
+    decision = 0.9
+
+    matches = matcher.get_index_pairs_within_thresholds(decision)
+    match_dict = dict(matches)
+
+    # value is added to "extracted_uid" column if match score is 1
+    cprr.loc[:, "matched_uid_8"] = cprr.uid_8.map(lambda x: match_dict.get(x))
+    return cprr
+
+
+def join_matched_uids(df):
+    df.loc[:, "post_id"] = df.matched_uid.fillna("").str.cat(
+        df.matched_uid_1.fillna(""), sep=","
+    )
+    df.loc[:, "post_id"] = df.post_id.str.cat(df.matched_uid_2.fillna(""), sep=",")
+    df.loc[:, "post_id"] = df.post_id.str.cat(df.matched_uid_3.fillna(""), sep=",")
+    df.loc[:, "post_id"] = df.post_id.str.cat(df.matched_uid_4.fillna(""), sep=",")
+    df.loc[:, "post_id"] = df.post_id.str.cat(df.matched_uid_5.fillna(""), sep=",")
+    df.loc[:, "post_id"] = df.post_id.str.cat(df.matched_uid_6.fillna(""), sep=",")
+    df.loc[:, "post_id"] = df.post_id.str.cat(df.matched_uid_7.fillna(""), sep=",")
+    df.loc[:, "post_id"] = df.post_id.str.cat(df.matched_uid_8.fillna(""), sep=",")
+
+    df.loc[:, "post_id"] = (
+        df.post_id.fillna("")
+        .str.replace(r"\,\, ?(.+)?", "", regex=True)
+        .str.replace(r"^\,", "", regex=True)
+    )
+    df = df.fillna("")
+    df = df[["post_id"]]
+    return df[~((df.post_id == ""))]
+
+
+if __name__ == "__main__":
+    cprr = pd.read_csv(deba.data("clean/post_officer_history.csv"))
+    pprr = pd.read_csv(deba.data("raw/fused/event.csv"))
```

Comment:
Why is the `event.csv` called `pprr`? Is this the same file as `fuse/event.csv`?

Reply:
It's poorly named. It's `fuse/event.csv` and `fuse/personnel.csv` merged into one table to match with `first_name`, `last_name`, and `agency`. I will correct.

Reply:
Okay, again don't put a file that was produced off of outputs of this pipeline into the raw folder. You're essentially cutting this file off of the pipeline so it will never be updated when either `fuse/event.csv` or `fuse/personnel.csv` is updated. And how this file came to be is also totally unknown. How can other people update it afterward?

Reply:
What is the best way to match these officers with those officers who exist in `fuse/personnel.csv`

Reply:
Simplest way: join the `first_name`, `last_name`, `middle_name` from `personnel` into `event` using `uid`. Then match using `ColumnsIndex(['fc', 'agency'])` and JaroWinkler on `first_name`, `last_name`, `middle_name`. Pretty straight-forward I think. Unless you have other columns that can also be used to match.

Reply:
On it. Thank you.

Reply:
Sorry for the misunderstanding. I think that I am still misunderstanding.

I was under the impression that at the `match` stage we should not be calling tables produced at the `fuse` stage, namely `fuse/personnel.csv` and `fuse/event.csv` because it would mess with the dependencies.

Reply:
You're right, It is such a bind but for now just make the new script a `fuse` script. Make all the scripts that depend on it a `fuse` script also. we can redistribute the scripts into existing stages or create new stages once we have the complete picture. Feel free to suggest new stages. I don't think `clean`, `match`, and `fuse` are sufficient anymore.

Reply:
After thinking about it some more. I think it's best to give Deba an option to disable stage order enforcing and instead make sure that Deba can detect circular dependency. That way you can have `match` script read from `fuse` output and vice versa but the whole pipeline would still form a DAG.

Reply:
Looks like Make already drop circular dependency by itself so all the changes I need to make is to add an option that disable stage order enforcement.

Reply:
Much better idea than what I was going to suggest. Awesome. I've been traveling a lot in the past few days, but I hope to have my changes to the script committed soon as well.

Reply:
I just pushed what I was going to suggest just for your general feedback. It is irrelevant now that you've decided the best solution, however, I wanted your feedback because while I think that it's not a great solution I can't exactly describe why. Would love your feedback. 

Still in the process of parameterizing and consolidating the match.

Reply:
@ayyubibrahimi Once Deba no longer enforce stage order, you should just put matching code in match scripts, cleaning code in clean scripts. etc. Putting all kinds of code in a single script also work but there are no structure, over time the project slowly descend into chaos. There always have to be some kind of structure in a project so that A) you know instantly where to put a piece of code and B) it becomes predictable and easy for others to maintain your code.

Since we cannot really structure this project order-wise, we have to structure it function-wise: organizing scripts based on their purpose. For now they will be `clean`, `match`, and `fuse` but in the future we will add `ocr` and `ner` as well.

Reply:
Makes perfect sense. Thank you.

Reply:
@ayyubibrahimi #283 should do it

Reply:
Awesome. Thanks for that.